### DISCUSSION TITLE
How can I use SSE with xstate?

### DISCUSSION THREAD

**Discussion by Kira-Pilot** [statelyai/xstate#3608] [Q&A] [answered]
Hi There!
I'm trying to use [SSE](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events) with xstate. I have a machine which has the following state:

````
watchingEvents: {
  invoke: {
    src: (context) => (callback) => {
      const sse = new EventSource(`${location.protocol}//${location.host}/events/watch`,
        { withCredentials: true }
      )

      // successfully logs "Do I make it in here?"
      console.log('Do I make it in here?')

      // the sse is not logged
      console.log('sse', sse)

      sse.onmessage = (e) => {
        // the event is not logged either
        console.log('e', e)

        // the event is not sent
        callback({ type: 'UPDATE_EVENT, e});
      }
    }
  }, 
````

The events are definitely present - I can watch them come in via the network tab in Chrome.

Also, events work as expected if I use vanilla React in my page:

````
useEffect(() => {

  const sse = new EventSource(
    `${location.protocol}//${location.host}/events/watch`,
    { withCredentials: true },
  )

 // this logs successfully
 sse.onmessage = (e) => console.log("event", e)

}, [])
````

I'm using xstate 4.33.5. Any idea where I'm going astray?

**Comment by Andarist** [comment, marked as answer]
```ts
      // successfully logs "Do I make it in here?"
      console.log('Do I make it in here?')

      // the sse is not logged
      console.log('sse', sse)
```

This part looks suspicious - I mean, this shouldn't be possible. Both of those console.logs should always be executed here, they are part of the same function and they are called synchronously.

The code itself looks OK at the first glance.

**Reply by Kira-Pilot** [reply]
@Andarist, sorry - that wasn't very clear. I was getting the second `console.log`, but sse was `undefined`. 

However, I think this is resolved. Looks like our API started sending labeled events, so I should have been using a listener. Thanks for the reply anyway, and I'll close this out.
